File: c/src/queue.c
#include "zeroipc.h"
#include "zeroipc_queue.h"
#include <stdatomic.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>

/*
 * Queue binary layout (matches C++/Go/Python):
 *   [head:u32][tail:u32][capacity:u32][elem_size:u32]  (16 bytes)
 *   [data[0]][data[1]]...[data[cap-1]]
 *   [seq[0]:u32][seq[1]:u32]...[seq[cap-1]:u32]
 *
 * Uses Vyukov bounded MPMC queue algorithm with per-slot sequence numbers.
 */

/* Queue header in shared memory — matches C++ Queue::Header */
typedef struct {
    _Atomic uint32_t head;
    _Atomic uint32_t tail;
    uint32_t capacity;
    uint32_t elem_size;
} queue_header_t;

/* Queue structure (process-local handle) */
struct zeroipc_queue {
    zeroipc_memory_t* memory;
    queue_header_t* header;
    void* data;                  /* pointer to data array */
    _Atomic uint32_t* seq;       /* pointer to sequence array */
    char name[32];
};

/* Create queue */
zeroipc_queue_t* zeroipc_queue_create(zeroipc_memory_t* mem, const char* name,
                                       size_t elem_size, size_t capacity) {
    if (!mem || !name || elem_size == 0 || capacity == 0) {
        return NULL;
    }

    zeroipc_queue_t* queue = calloc(1, sizeof(zeroipc_queue_t));
    if (!queue) return NULL;

    queue->memory = mem;
    strncpy(queue->name, name, sizeof(queue->name) - 1);

    /* Layout: [header][data: elem_size * capacity][seq: uint32 * capacity] */
    size_t seq_array_size = sizeof(uint32_t) * capacity;
    if (capacity > (SIZE_MAX - sizeof(queue_header_t) - seq_array_size) / elem_size) {
        free(queue);
        return NULL;
    }
    size_t total_size = sizeof(queue_header_t) + elem_size * capacity + seq_array_size;

    size_t offset;
    int result = zeroipc_table_add(mem, name, total_size, &offset);
    if (result != ZEROIPC_OK) {
        free(queue);
        return NULL;
    }

    queue->header = (queue_header_t*)((char*)zeroipc_memory_base(mem) + offset);
    queue->data = (char*)queue->header + sizeof(queue_header_t);
    queue->seq = (_Atomic uint32_t*)((char*)queue->data + elem_size * capacity);

    /* Initialize header */
    atomic_store(&queue->header->head, 0);
    atomic_store(&queue->header->tail, 0);
    queue->header->capacity = capacity;
    queue->header->elem_size = elem_size;

    /* Initialize per-slot sequence numbers: seq[i] = i */
    for (uint32_t i = 0; i < capacity; ++i) {
        atomic_store_explicit(&queue->seq[i], i, memory_order_relaxed);
    }

    return queue;
}

/* Open existing queue */
zeroipc_queue_t* zeroipc_queue_open(zeroipc_memory_t* mem, const char* name) {
    if (!mem || !name) return NULL;

    size_t offset, size;
    int result = zeroipc_table_find(mem, name, &offset, &size);
    if (result != ZEROIPC_OK) return NULL;

    zeroipc_queue_t* queue = calloc(1, sizeof(zeroipc_queue_t));
    if (!queue) return NULL;

    queue->memory = mem;
    strncpy(queue->name, name, sizeof(queue->name) - 1);

    queue->header = (queue_header_t*)((char*)zeroipc_memory_base(mem) + offset);
    if (queue->header->capacity == 0 || queue->header->elem_size == 0) {
        free(queue);
        return NULL;
    }

    queue->data = (char*)queue->header + sizeof(queue_header_t);
    queue->seq = (_Atomic uint32_t*)(
        (char*)queue->data + queue->header->elem_size * queue->header->capacity);

    return queue;
}

void zeroipc_queue_close(zeroipc_queue_t* queue) {
    free(queue);
}

/* Push (Vyukov bounded MPMC) */
int zeroipc_queue_push(zeroipc_queue_t* queue, const void* value) {
    if (!queue || !value) return ZEROIPC_ERROR_SIZE;

    uint32_t cap = queue->header->capacity;

    for (;;) {
        uint32_t tail = atomic_load_explicit(&queue->header->tail, memory_order_relaxed);
        uint32_t slot = tail % cap;
        uint32_t s = atomic_load_explicit(&queue->seq[slot], memory_order_acquire);
        int32_t diff = (int32_t)s - (int32_t)tail;

        if (diff == 0) {
            if (atomic_compare_exchange_weak_explicit(
                    &queue->header->tail, &tail, tail + 1,
                    memory_order_relaxed, memory_order_relaxed)) {
                void* dest = (char*)queue->data + slot * queue->header->elem_size;
                memcpy(dest, value, queue->header->elem_size);
                atomic_store_explicit(&queue->seq[slot], tail + 1, memory_order_release);
                return ZEROIPC_OK;
            }
        } else if (diff < 0) {
            return ZEROIPC_ERROR_SIZE;  /* full */
        }
    }
}

/* Pop (Vyukov bounded MPMC) */
int zeroipc_queue_pop(zeroipc_queue_t* queue, void* value) {
    if (!queue || !value) return ZEROIPC_ERROR_SIZE;

    uint32_t cap = queue->header->capacity;

    for (;;) {
        uint32_t head = atomic_load_explicit(&queue->header->head, memory_order_relaxed);
        uint32_t slot = head % cap;
        uint32_t s = atomic_load_explicit(&queue->seq[slot], memory_order_acquire);
        int32_t diff = (int32_t)s - (int32_t)(head + 1);

        if (diff == 0) {
            if (atomic_compare_exchange_weak_explicit(
                    &queue->header->head, &head, head + 1,
                    memory_order_relaxed, memory_order_relaxed)) {
                void* src = (char*)queue->data + slot * queue->header->elem_size;
                memcpy(value, src, queue->header->elem_size);
                atomic_store_explicit(&queue->seq[slot], head + cap, memory_order_release);
                return ZEROIPC_OK;
            }
        } else if (diff < 0) {
            return ZEROIPC_ERROR_NOT_FOUND;  /* empty */
        }
    }
}

int zeroipc_queue_empty(zeroipc_queue_t* queue) {
    if (!queue) return 1;
    uint32_t head = atomic_load_explicit(&queue->header->head, memory_order_acquire);
    uint32_t tail = atomic_load_explicit(&queue->header->tail, memory_order_acquire);
    return head == tail;
}

int zeroipc_queue_full(zeroipc_queue_t* queue) {
    if (!queue) return 1;
    uint32_t head = atomic_load_explicit(&queue->header->head, memory_order_acquire);
    uint32_t tail = atomic_load_explicit(&queue->header->tail, memory_order_acquire);
    return (tail - head) >= queue->header->capacity;
}

size_t zeroipc_queue_size(zeroipc_queue_t* queue) {
    if (!queue) return 0;
    uint32_t head = atomic_load_explicit(&queue->header->head, memory_order_acquire);
    uint32_t tail = atomic_load_explicit(&queue->header->tail, memory_order_acquire);
    /* uint32_t subtraction handles wraparound correctly */
    return (size_t)(tail - head);
}

size_t zeroipc_queue_capacity(zeroipc_queue_t* queue) {
    return queue ? queue->header->capacity : 0;
}

File: go/zeroipc/queue.go
package zeroipc

import (
	"encoding/binary"
	"errors"
	"fmt"
	"sync/atomic"
	"unsafe"
)

// QueueHeaderSize is the size of the queue header in bytes.
// Layout: head(4) + tail(4) + capacity(4) + elem_size(4) = 16 bytes
const QueueHeaderSize = 16

// Queue is a lock-free MPMC (multi-producer multi-consumer) circular buffer
// using the Vyukov bounded queue algorithm with per-slot sequence numbers.
//
// Binary layout (matching C++):
//   - head: uint32 (atomic, offset 0)
//   - tail: uint32 (atomic, offset 4)
//   - capacity: uint32 (offset 8)
//   - elem_size: uint32 (offset 12)
//   - data: capacity * elem_size bytes (offset 16)
//   - sequence: capacity * 4 bytes (after data) - per-slot sequence numbers
type Queue[T Numeric] struct {
	memory   *Memory
	name     string
	offset   int
	capacity uint32
	elemSize uint32
}

// NewQueue creates a new queue in shared memory.
func NewQueue[T Numeric](memory *Memory, name string, capacity int) (*Queue[T], error) {
	if len(name) >= NameSize {
		return nil, errors.New("name too long (max 31 characters)")
	}

	if capacity <= 0 {
		return nil, errors.New("capacity must be positive")
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	// Check if already exists
	if entry := memory.Find(name); entry != nil {
		return nil, fmt.Errorf("queue '%s' already exists", name)
	}

	// Layout: [Header(16)][data: T*capacity][pad][sequence: uint32*capacity]
	totalSize := QueueHeaderSize + align8(capacity*elemSize) + capacity*4
	offset, err := memory.Allocate(name, totalSize)
	if err != nil {
		return nil, fmt.Errorf("failed to allocate: %w", err)
	}

	data := memory.Data()

	// Write header
	binary.LittleEndian.PutUint32(data[offset:], 0)                    // head
	binary.LittleEndian.PutUint32(data[offset+4:], 0)                  // tail
	binary.LittleEndian.PutUint32(data[offset+8:], uint32(capacity))   // capacity
	binary.LittleEndian.PutUint32(data[offset+12:], uint32(elemSize))  // elem_size

	// Zero-initialize data area
	dataStart := offset + QueueHeaderSize
	dataEnd := dataStart + capacity*elemSize
	for i := dataStart; i < dataEnd; i++ {
		data[i] = 0
	}

	// Initialize per-slot sequence numbers: sequence[i] = i (8-aligned start)
	seqStart := dataStart + align8(capacity*elemSize)
	for i := 0; i < capacity; i++ {
		seqOffset := seqStart + i*4
		binary.LittleEndian.PutUint32(data[seqOffset:], uint32(i))
	}

	return &Queue[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: uint32(capacity),
		elemSize: uint32(elemSize),
	}, nil
}

// OpenQueue opens an existing queue in shared memory.
func OpenQueue[T Numeric](memory *Memory, name string) (*Queue[T], error) {
	entry := memory.Find(name)
	if entry == nil {
		return nil, fmt.Errorf("queue '%s' not found", name)
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	offset := int(entry.Offset)
	data := memory.Data()

	// Read header
	capacity := binary.LittleEndian.Uint32(data[offset+8:])
	storedElemSize := binary.LittleEndian.Uint32(data[offset+12:])

	if int(storedElemSize) != elemSize {
		return nil, fmt.Errorf("element size mismatch: stored %d, expected %d", storedElemSize, elemSize)
	}

	return &Queue[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: capacity,
		elemSize: uint32(elemSize),
	}, nil
}

// headPtr returns a pointer to the atomic head counter.
func (q *Queue[T]) headPtr() *uint32 {
	return (*uint32)(unsafe.Pointer(&q.memory.Data()[q.offset]))
}

// tailPtr returns a pointer to the atomic tail counter.
func (q *Queue[T]) tailPtr() *uint32 {
	return (*uint32)(unsafe.Pointer(&q.memory.Data()[q.offset+4]))
}

// seqPtr returns a pointer to the sequence number for slot i.
// Sequence array lives after the data array.
func (q *Queue[T]) seqPtr(slot uint32) *uint32 {
	seqOffset := q.offset + QueueHeaderSize + align8(int(q.capacity)*int(q.elemSize)) + int(slot)*4
	return (*uint32)(unsafe.Pointer(&q.memory.Data()[seqOffset]))
}

// Push adds an element to the queue (lock-free Vyukov MPMC).
// Returns false if the queue is full.
func (q *Queue[T]) Push(value T) bool {
	cap := q.capacity
	tailPtr := q.tailPtr()

	for {
		tail := atomic.LoadUint32(tailPtr)
		slot := tail % cap
		seq := atomic.LoadUint32(q.seqPtr(slot))
		diff := int32(seq) - int32(tail)

		if diff == 0 {
			// Slot is ready for writing; try to claim it
			if atomic.CompareAndSwapUint32(tailPtr, tail, tail+1) {
				// We own this slot — write the data
				dataOffset := q.offset + QueueHeaderSize + int(slot)*int(q.elemSize)
				*(*T)(unsafe.Pointer(&q.memory.Data()[dataOffset])) = value
				// Publish: set sequence to tail+1 so consumers can see it
				atomic.StoreUint32(q.seqPtr(slot), tail+1)
				return true
			}
			// CAS failed, another producer got it; retry
		} else if diff < 0 {
			// Queue is full
			return false
		}
		// diff > 0: stale tail or another producer in progress; retry
	}
}

// Pop removes and returns an element from the queue (lock-free Vyukov MPMC).
// Returns the value and true if successful, or zero value and false if empty.
func (q *Queue[T]) Pop() (T, bool) {
	var zero T
	cap := q.capacity
	headPtr := q.headPtr()

	for {
		head := atomic.LoadUint32(headPtr)
		slot := head % cap
		seq := atomic.LoadUint32(q.seqPtr(slot))
		diff := int32(seq) - int32(head+1)

		if diff == 0 {
			// Slot contains data; try to claim it
			if atomic.CompareAndSwapUint32(headPtr, head, head+1) {
				// We own this slot — read the data
				dataOffset := q.offset + QueueHeaderSize + int(slot)*int(q.elemSize)
				value := *(*T)(unsafe.Pointer(&q.memory.Data()[dataOffset]))
				// Release: set sequence to head+capacity so producers can reuse
				atomic.StoreUint32(q.seqPtr(slot), head+cap)
				return value, true
			}
			// CAS failed, another consumer got it; retry
		} else if diff < 0 {
			// Queue is empty
			return zero, false
		}
		// diff > 0: stale head or another consumer in progress; retry
	}
}

// TryPush is an alias for Push (both are non-blocking).
func (q *Queue[T]) TryPush(value T) bool {
	return q.Push(value)
}

// TryPop is an alias for Pop (both are non-blocking).
func (q *Queue[T]) TryPop() (T, bool) {
	return q.Pop()
}

// Empty returns true if the queue appears empty.
// Note: In a concurrent context, this is only an approximation.
func (q *Queue[T]) Empty() bool {
	return atomic.LoadUint32(q.headPtr()) == atomic.LoadUint32(q.tailPtr())
}

// Full returns true if the queue appears full.
// Note: In a concurrent context, this is only an approximation.
func (q *Queue[T]) Full() bool {
	head := atomic.LoadUint32(q.headPtr())
	tail := atomic.LoadUint32(q.tailPtr())
	return (tail - head) >= q.capacity
}

// Size returns the approximate number of elements in the queue.
// Note: In a concurrent context, this is only an approximation.
func (q *Queue[T]) Size() int {
	head := atomic.LoadUint32(q.headPtr())
	tail := atomic.LoadUint32(q.tailPtr())
	// uint32 subtraction handles wraparound correctly
	return int(tail - head)
}

// Capacity returns the maximum number of elements the queue can hold.
func (q *Queue[T]) Capacity() int {
	return int(q.capacity)
}

// Name returns the queue name.
func (q *Queue[T]) Name() string {
	return q.name
}
